src/Compilers/Core/Portable/SourceGeneration/Nodes/TransformNode.cs

```diff
@@ -0,0 +1,71 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Text;
+using System.Threading;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal class TransformNode<TInput, TOutput> : IIncrementalGeneratorNode<TOutput>
+    {
+        private readonly Func<TInput, IEnumerable<TOutput>> _func;
+        private readonly IIncrementalGeneratorNode<TInput> _sourceNode;
+
+        public TransformNode(IIncrementalGeneratorNode<TInput> sourceNode, Func<TInput, TOutput> userFunc)
+            : this(sourceNode, userFunc: (i) => ImmutableArray.Create(userFunc(i)))
+        {
+        }
+
+        public TransformNode(IIncrementalGeneratorNode<TInput> sourceNode, Func<TInput, IEnumerable<TOutput>> userFunc)
+        {
+            _sourceNode = sourceNode;
+            _func = userFunc;
+        }
+
+        // PROTOTYPE(source-generators):
+        public IIncrementalGeneratorNode<TOutput> WithComparer(IEqualityComparer<TOutput> comparer) => this;
+
+        public NodeStateTable<TOutput> UpdateStateTable(DriverStateTable.Builder builder, NodeStateTable<TOutput> previousTable, CancellationToken cancellationToken)
+        {
+            // PROTOTYPE(source-generators):caching, faulted etc.
+
+            // Semantics of a transform:
+            // Element-wise comparison of upstream table
+            // - Cached or Removed: no transform, just use previous values
+            // - Added: perform transform and add
+            // - Modified: perform transform and do element wise comparison with previous results
+
+            // grab the source inputs
+            var sourceTable = builder.GetLatestStateTableForNode(_sourceNode);
+            var newTable = new NodeStateTable<TOutput>.Builder();
+
+            foreach (var entry in sourceTable)
+            {
+                if (entry.state == EntryState.Cached || entry.state == EntryState.Removed)
+                {
+                    newTable.AddEntriesFromPreviousTable(previousTable, entry.state);
+                }
+                else
+                {
+                    // generate the new entries
+                    var newOutputs = _func(entry.item).ToImmutableArray();
+
+                    if (entry.state == EntryState.Added)
+                    {
+                        newTable.AddEntries(newOutputs, EntryState.Added);
+                    }
+                    else
+                    {
+                        newTable.ModifyEntriesFromPreviousTable(previousTable, newOutputs);
```

Comment:
Reading through `ModifyEntriesFromPreviousTable` I believe their is an assumption that `newOutputs` here is order dependent. Basically that the source generator providing `_func` to this node has an obligation to produce the same output in the same order for the same input. 

Is my understanding correct here and if so where do we document that? 

Most importantly if the generator violates that what goes wrong? Do we gracefully recover somehow?

Reply:
Yes, it's written in the design doc that we require transforms to be deterministic (both for ordering, and because we'll cache in many cases and not call again). If an author is deliberatly non determisinstc, we'll just end up with a 'polluted' state table, where we think things are modified that weren't. That won't have any correctness concerns, but will cause the downstream nodes to run unneccasarily, potentially hurting perf. TL;DR: determinism is needed for perf, but not corectness.

---
In reply to: [619517291](https://github.com/dotnet/roslyn/pull/52809#discussion_r619517291) [](ancestors = 619517291)